packages/contracts/src/EmailAuth.sol

```diff
@@ -0,0 +1,182 @@
+
+            if(Strings.equal(template[i], "{string}")) {
+                string memory param = abi.decode(emailAuthMsg.subjectParams[nextParamIndex], (string));
+                stringParam = param;
+                // expectedSubject = string(abi.encodePacked(expectedSubject, param));
+            } else if(Strings.equal(template[i], "{uint}")) {
+                uint256 param = abi.decode(emailAuthMsg.subjectParams[nextParamIndex], (uint256));
+                stringParam = Strings.toString(param);
+                // expectedSubject = string(abi.encodePacked(expectedSubject, Strings.toString(param)));
+            } else if(Strings.equal(template[i], "{int}")) {
+                int256 param = abi.decode(emailAuthMsg.subjectParams[nextParamIndex], (int256));
+                stringParam = Strings.toString(param);
+                // expectedSubject = string(abi.encodePacked(expectedSubject, Strings.toString(param)));
+            } else if(Strings.equal(template[i], "{decimals}")) {
+                uint256 param = abi.decode(emailAuthMsg.subjectParams[nextParamIndex], (uint256));
+                stringParam = Strings.toString(param);
+                // expectedSubject = string(abi.encodePacked(expectedSubject, Strings.toString(param)));
+            } else if(Strings.equal(template[i], "{decimals}")) {
+                uint256 param = abi.decode(emailAuthMsg.subjectParams[nextParamIndex], (uint256));
+                stringParam = Strings.toString(param);
+                // expectedSubject = string(abi.encodePacked(expectedSubject, Strings.toString(param)));
+            } else if(Strings.equal(template[i], "{ethAddr}")) {
+                address param = abi.decode(emailAuthMsg.subjectParams[nextParamIndex], (address));
+                stringParam = Strings.toHexString(param);
+            } else {
+                continue;
+            }
+
+            if(i > 0) {
+                expectedSubject = string(abi.encodePacked(expectedSubject, " "));
+            }
+            expectedSubject = string(abi.encodePacked(expectedSubject, stringParam));
+            nextParamIndex++;
+        }
+        string memory trimmedSubject = removePrefix(emailAuthMsg.proof.maskedSubject, emailAuthMsg.skipedSubjectPrefix);
+        require(Strings.equal(expectedSubject, trimmedSubject), "invalid subject");
+        require(verifier.verifyEmailProof(emailAuthMsg.proof) == true, "invalid email proof");
+
+        bytes32 msgHash = computeMsgHash(
+            emailAuthMsg.proof.accountSalt, 
+            emailAuthMsg.proof.isCodeExist, 
+            emailAuthMsg.templateId, 
+            emailAuthMsg.subjectParams
+        );
+
+        require(authedHash[emailAuthMsg.proof.emailNullifier] == bytes32(0), "email already authed");
+        authedHash[emailAuthMsg.proof.emailNullifier] = msgHash;
+
+        return msgHash;
+    }
+
+    function isValidSignature(bytes32 _hash, bytes memory _signature) public view returns (bytes4) {
```

Comment:
I didn't use _hash of arguments. Is that acceptable?

Reply:
Sorry, my spec is wrong.
1. _signature is just an encoding of the `emailNullifier`.
2. This function does not need to compute `msgHash`.
3. It checks if `authedHash[emailNullifier]==_hash`.